api/handlers.go

```diff
@@ -447,25 +463,73 @@ func (api *API) userLimitsGET(_ *database.User, w http.ResponseWriter, req *http
 			api.WriteJSON(w, respAnon)
 			return
 		}
-		api.staticUserTierCache.Set(u)
+		api.staticUserTierCache.Set(u.Sub, u)
 		// Populate the tier and qe values, while simultaneously making sure
 		// that we can read the record from the cache.
-		tier, qe, ok = api.staticUserTierCache.Get(sub)
+		tier, qe, ok = api.staticUserTierCache.Get(u.Sub)
 		if !ok {
 			build.Critical("Failed to fetch user from UserTierCache right after setting it.")
 		}
 	}
-	resp := userLimitsGetFromTier(tier)
-	// If the quota is exceeded we should keep the user's tier but report
-	// anonymous-level speeds.
-	if qe {
-		// Report anonymous speeds.
-		resp = userLimitsGetFromTier(database.TierAnonymous)
-		// Keep reporting the user's actual tier and tier name.
-		resp.TierID = tier
-		resp.TierName = database.UserLimits[tier].TierName
+	api.WriteJSON(w, userLimitsGetFromTier(tier, qe))
+}
+
+// userLimitsSkylinkGET returns the speed limits which apply to a GET call to
+// the given skylink. This method exists to accommodate public API keys.
+//
+// NOTE: This handler needs to use the noAuth middleware in order to be able to
+// optimise its calls to the DB and the use of caching.
+func (api *API) userLimitsSkylinkGET(u *database.User, w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	respAnon := userLimitsGetFromTier(database.TierAnonymous, false)
+	// Validate the skylink.
+	skylink := ps.ByName("skylink")
+	if !database.ValidSkylinkHash(skylink) {
+		api.staticLogger.Tracef("Invalid skylink: %s", skylink)
+		api.WriteJSON(w, respAnon)
+		return
 	}
-	api.WriteJSON(w, resp)
+	// Try to fetch an API attached to the request.
+	ak, err := apiKeyFromRequest(req)
+	if errors.Contains(err, ErrNoAPIKey) {
+		// We failed to fetch an API key from this request but the request might
+		// be authenticated in another way, so we'll defer to userLimitsGET.
+		api.userLimitsGET(u, w, req, ps)
+		return
+	}
+	if err != nil {
+		api.staticLogger.Debugf("Error while processing API key: %s", err)
+		api.WriteJSON(w, respAnon)
+		return
+	}
+	// Check the cache before hitting the database.
+	tier, qe, ok := api.staticUserTierCache.Get(ak.String() + skylink)
```

Comment:
`nit` ok -> cached

Reply:
I prefer to keep the `ok`, if that's fine with you.